File: Medium/Combination Sum.cpp
//https://leetcode.com/explore/challenge/card/october-leetcoding-challenge/559/week-1-october-1st-october-7th/3481/
class Solution {
public:
    
    void dfs(int j, int sum, vector<int> & take, vector<vector<int>> & ans, int target, vector<int> candidates) {
        if(sum-target== 0){
            ans.push_back(take);
            take.pop_back();
            return;
        }
        if(sum-target > 0) {
           take.pop_bac
               k();
            return;
        }
        for(int i = j; i < candidates.size(); i++) {
            take.push_back(candidates[i]);
            dfs(i, sum+candidates[i], take, ans, target, candidates);
            
        }
        take.pop_back();
    }
     vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<int> take;
        vector<vector<int>> ans;
        int sum =0;
        dfs(0, sum, take, ans, target, candidates);
        return ans;
    }
};

File: 39-combination-sum/combination-sum.java
class Solution {
    public void backtrack(int[] candidates, List<List<Integer>> list, List<Integer> temp, int start, int target) {
        if (target == 0) {
            list.add(new ArrayList<>(temp));
        }

        if (target < 0) {
            return;
        }

        for (int i = start; i < candidates.length; i++) {
            temp.add(candidates[i]);
            backtrack(candidates, list, temp, i, target - candidates[i]);
            temp.remove(temp.size()-1);
        }
    }
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        
        List<List<Integer>> list = new ArrayList<>();
        backtrack(candidates, list, new ArrayList<>(), 0, target);
        return list;
    }
}